apps/omg/lib/omg/state/core.ex

```diff
@@ -99,31 +99,24 @@ defmodule OMG.State.Core do
   Recovers the ledger's state from data delivered by the `OMG.DB`
   """
   @spec extract_initial_state(
-          utxos_query_result :: [list({OMG.DB.utxo_pos_db_t(), OMG.Utxo.t()})],
           height_query_result :: non_neg_integer() | :not_found,
           child_block_interval :: pos_integer()
         ) :: {:ok, t()} | {:error, :top_block_number_not_found}
-  def extract_initial_state(utxos_query_result, height_query_result, child_block_interval)
-      when is_list(utxos_query_result) and is_integer(height_query_result) and is_integer(child_block_interval) do
-    # extract height, last deposit height and utxos from query result
-    height = height_query_result + child_block_interval
-
-    state = %__MODULE__{
-      height: height,
-      utxos: UtxoSet.init(utxos_query_result)
-    }
+  def extract_initial_state(height_query_result, child_block_interval)
+      when is_integer(height_query_result) and is_integer(child_block_interval) do
+    state = %__MODULE__{height: height_query_result + child_block_interval}
 
     {:ok, state}
   end
 
-  def extract_initial_state(
-        _utxos_query_result,
-        :not_found,
-        _child_block_interval
-      ) do
+  def extract_initial_state(:not_found, _child_block_interval) do
     {:error, :top_block_number_not_found}
   end
 
+  @spec with_utxos(t(), UtxoSet.query_result_t()) :: t()
+  def with_utxos(%Core{utxos: utxos} = state, utxos_query_result),
+    do: %{state | utxos: UtxoSet.merge_with_query_result(utxos, utxos_query_result)}
```

Comment:
ah, there's a more serious problem I missed in my experiments. If there are 2 txs that spend the same UTXO in a single block, they'll pass (there should be appropriate tests for this being impossible BTW, so the `State.CoreTest` should be expanded a lot). The first spends the UTXOs, but the second will pull them back from the DB again.

Looks like one needs to commit the `utxo_db_updates` to the DB on every `exec` in this case too.

Reply:
Ah, no - we couldn't commit the updates because if everything dies & restart before we form the block the state is corrupted.

Reply:
:bulb: Good idea: add test to show :point_up: problem.